src/expr/node_value.h

```diff
@@ -281,87 +221,144 @@ class NodeValue {
-    static NodeValue* s_null = new NodeValue(0);
-    return *s_null;
-  }
-
-  /**
-   * If this is a CONST_* Node, extract the constant from it.
-   */
-  template <class T>
-  inline const T& getConst() const;
+  std::string toString() const;
 
-  NodeValue* getOperator() const;
-  NodeValue* getChild(int i) const;
+  void toStream(std::ostream& out,
+                int toDepth = -1,
+                bool types = false,
+                size_t dag = 1,
+                OutputLanguage = language::output::LANG_AUTO) const;
 
   void printAst(std::ostream& out, int indent = 0) const;
 
-private:
+  template <typename T>
+  inline iterator<T> begin() const;
+  template <typename T>
+  inline iterator<T> end() const;
+
+  /* ------------------------------------------------------------------------ */
+ private:
+  /* ------------------------------------------------------------------------ */
 
   /**
    * RAII guard that increases the reference count if the reference count to be
    * > 0.  Otherwise, this does nothing. This does not just increment the
    * reference count to avoid maxing out the d_rc field. This is only for low
    * level functions.
    */
-  class RefCountGuard {
-    NodeValue* d_nv;
-    bool d_increased;
-  public:
-    RefCountGuard(const NodeValue* nv) :
-      d_nv(const_cast<NodeValue*>(nv)) {
+  class RefCountGuard
+  {
+   public:
+    RefCountGuard(const NodeValue* nv) : d_nv(const_cast<NodeValue*>(nv))
+    {
       d_increased = (d_nv->d_rc == 0);
-      if(d_increased) {
+      if (d_increased)
+      {
         d_nv->d_rc = 1;
       }
     }
-    ~RefCountGuard() {
+    ~RefCountGuard()
+    {
       // dec() without marking for deletion: we don't want to garbage
       // collect this NodeValue if ours is the last reference to it.
       // E.g., this can happen when debugging code calls the print
       // routines below.  As RefCountGuards are scoped on the stack,
       // this should be fine---but not in multithreaded contexts!
-      if(d_increased) {
+      if (d_increased)
+      {
         --d_nv->d_rc;
       }
     }
-  };/* NodeValue::RefCountGuard */
 
-  friend class RefCountGuard;
+   private:
+    NodeValue* d_nv;
+    bool d_increased;
+  }; /* NodeValue::RefCountGuard */
+
+  /* ------------------------------ Header ---------------------------------- */
+  static const uint32_t NBITS_REFCOUNT =
+      CVC4__EXPR__NODE_VALUE__NBITS__REFCOUNT;
+  static const uint32_t NBITS_KIND = CVC4__EXPR__NODE_VALUE__NBITS__KIND;
+  static const uint32_t NBITS_ID = CVC4__EXPR__NODE_VALUE__NBITS__ID;
+  static const uint32_t NBITS_NCHILDREN =
+      CVC4__EXPR__NODE_VALUE__NBITS__NCHILDREN;
+
+  /** Maximum reference count possible.  Used for sticky
+   *  reference-counting.  Should be (1 << num_bits(d_rc)) - 1 */
+  static const uint32_t MAX_RC = (((uint32_t)1) << NBITS_REFCOUNT) - 1;
+
+  /** A mask for d_kind */
+  static const uint32_t kindMask = (((uint32_t)1) << NBITS_KIND) - 1;
```

Comment:
Data members actually go to the end of the class according to our ordering.

Reply:
Good catch! I did the reordering twice in the end since I had to narrow down compilation issues and this slipped through somehow.

Reply:
Our guidelines say that this does not apply to constants, they come after types and nested structs and classes: https://github.com/CVC4/CVC4/wiki/Code-Style-Guidelines#declaration-order

Reply:
Ah true, I was focusing on them being static data members.